### PR TITLE
[citycar] can specify the sparsity as a probability

### PR DESCRIPTION
**Pull request by guicho271828**

### CODE REVIEW

File: citycar/generator.py [hunk 1]
```diff
@@ -7,36 +7,31 @@
...
+garage = args.garages
+sparse = args.sparse
+seed   = args.seed
 
-if len(sys.argv) == 7:
-    seed = int(sys.argv[6])
-else:
-    seed = None
+assert 0.0 <= sparse, "sparsity must be a probability (0.0 <= sparse <= 1.0)"
+assert sparse <= 1.0, "sparsity must be a probability (0.0 <= sparse <= 1.0)"
```
Comment (jendrikseipp) [inline citycar/generator.py:34]: I'd use `parser.error()` for user-facing messages about incorrect parameters (https://stackoverflow.com/questions/8107713/using-argparse-argumenterror-in-python).

**Review by jendrikseipp — CHANGES_REQUESTED** [review]

File: citycar/generator.py [hunk 2]
```diff
@@ -7,36 +7,31 @@
...
+to_print_car_pos = ""
 
+parser = argparse.ArgumentParser()
 
-to_print_car_pos = ""
+parser.add_argument("rows",type=int,"the grid row.")
+parser.add_argument("columns",type=int,"the grid column.")
+parser.add_argument("cars",type=int,"how many cars have to go through the network")
+parser.add_argument("garages",type=int,"the number of starting garages")
+parser.add_argument("--sparse",type=float,default=1.0,"The ratio of the available roads.")
```
[the remark this answers is not in the document]
Comment (guicho271828) [reply citycar/generator.py:21]: Maybe we should rename it? --damage-ratio, --avail-ratio, --availability ...

[the remark this answers is not in the document]
Comment (jendrikseipp) [reply citycar/generator.py:21]: What about "density"?

[the remark this answers is not in the document]
Comment (guicho271828) [reply citycar/generator.py]: done

File: citycar/generator.py [hunk 3]
```diff
@@ -7,36 +7,31 @@
...
-    )
-    print("\t seed is the (optional) random seed for the instance")
-    sys.exit(2)
+to_print_car_pos = ""
 
+parser = argparse.ArgumentParser()
 
-to_print_car_pos = ""
+parser.add_argument("rows",type=int,"the grid row.")
+parser.add_argument("columns",type=int,"the grid column.")
```
[the remark this answers is not in the document]
Comment (guicho271828) [reply citycar/generator.py]: done

File: citycar/generator.py (hunk 1 above)
Comment (guicho271828) [reply citycar/generator.py]: done

File: citycar/generator.py [hunk 4]
```diff
@@ -7,36 +7,32 @@
...
+garage  = args.garages
+density = args.density
+seed    = args.seed
 
-row = int(sys.argv[1])
-column = int(sys.argv[2])
-car = int(sys.argv[3])
-garage = int(sys.argv[4])
-sparse = int(sys.argv[5])
+if density < 0.0 or 1.0 < density:
```
Comment (jendrikseipp) [inline citycar/generator.py:33]: We could use `if not (0.0 <= density <= 1.0):`

File: citycar/README [hunk 5]
```diff
@@ -18,3 +18,14 @@ built for connecting two junctions and allowing a car to move between
...
+
+
+
+pddl-generator note:
+
+Sparsity was a 0 / 1 flag in the original version.
+When it is enabled, 20% of the roads become unavailable.
+
+In this version, we can specify the sparsity as a probability (default: 1.0).
+To achieve the same sparsity as the original, give 0.8 as a parameter.
```
Comment (jendrikseipp) [inline citycar/README:30]: Could you update this text for the new parameter name "density"?

**Review by jendrikseipp — CHANGES_REQUESTED** [review]

File: citycar/generator.py (hunk 4 above)
Comment (guicho271828) [reply citycar/generator.py:33]: oh well, what kind of method override trick python is using to allow it to be a valid code... I really hate python.
All languages should be like
> (<= 0.0 density 1.0)

File: citycar/generator.py [hunk 6]
```diff
@@ -190,9 +186,8 @@ def help():
 # free cells
 for i in range(row):
     for j in range(column):
-        if sparse != 0 and i != 0 and i < row - 1 and j != 0 and j < column - 1:
-            x = random.randint(0, 10)
-            if x < 8:
+        if density != 1.0 and i != 0 and i < row - 1 and j != 0 and j < column - 1:
+            if density < random.random():
```
[the remark this answers is not in the document]
Comment (guicho271828) [reply citycar/generator.py]: thanks for catching this!

**Comment by jendrikseipp** [comment]
Thanks!

**Pull request merged**